library/data/hash_map.lean

```diff
@@ -65,34 +124,597 @@ def erase_aux (a : α) : list (Σ a, β a) → list (Σ a, β a)
 | []            := []
 | (⟨a', b'⟩::t) := if a' = a then t else ⟨a', b'⟩ :: erase_aux t
 
-def insert (m : hash_map α β) (a : α) (b : β a) : hash_map α β :=
-match m with
-| ⟨hash_fn, size, nbuckets, nz, buckets⟩ :=
-  let bidx := mk_idx nz (hash_fn a) in
-  let bkt  := buckets.read bidx in
-  if contains_aux a bkt
-  then ⟨hash_fn, size, nbuckets, nz, buckets.write bidx (replace_aux a b bkt)⟩
-  else let size'    := size + 1 in
-       let buckets' := buckets.write bidx (⟨a, b⟩::bkt) in
-       if size' <= nbuckets
-       then ⟨hash_fn, size', nbuckets, nz, buckets'⟩
-       else let nbuckets' := nbuckets * 2 in
-            let nz' : nbuckets' > 0 := mul_pos nz dec_trivial in
-            ⟨hash_fn, size', nbuckets', nz',
-            fold_buckets buckets' (mk_array nbuckets' []) $
-               λ r a b, reinsert_aux hash_fn nz' r a b⟩
+inductive valid_aux {α : Type u} {β : α → Type v} (idx : α → nat) : Π (l : list (list Σ a, β a)) (sz : nat), Prop
+| nil {} : valid_aux [] 0
+| cons : Π (c : list Σ a, β a) {l sz}, valid_aux l sz → (c.map sigma.fst).nodup →
+  (∀ a ∈ c, idx (sigma.fst a) = l.length) → valid_aux (c::l) (sz + c.length)
+
+def valid_aux.unfold_cons {idx : α → nat} : Π {c l sz}, valid_aux idx (c::l) sz →
```

Comment:
The name "theorem" conveys a sense of significance, that's it.  Implementation-wise, they are virtually the same since they are token aliases: the only component of Lean that ever sees the difference is the scanner, for both lemma and theorem it returns the `theorem` token.